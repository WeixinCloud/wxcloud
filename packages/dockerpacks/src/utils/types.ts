export type NonEmptyArray<T> = [T, ...T[]];

export function isGenericNonEmptyArray(input: any): input is NonEmptyArray<unknown> {
  return Array.isArray(input) && isNonEmptyArray(input);
}

export function isNonEmptyArray<T = unknown>(input: Array<T>): input is NonEmptyArray<T> {
  return input.length > 0;
}
