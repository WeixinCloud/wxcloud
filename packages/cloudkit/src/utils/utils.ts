/**
 * isMatchMajorVersion('^1.2.3', '1') -> true
 *
 * support ~version & ^version
 * @param version currentVersion
 * @param majorVersion target majorVersion
 */
export const isMatchMajorVersion = (version: string, majorVersion: string) => {
  return (
    version.startsWith(`^${majorVersion}`) ||
    version.startsWith(`~${majorVersion}`) ||
    version.startsWith(majorVersion)
  );
};

/**
 * safely get deps value from packageJson
 * @param packageJson
 * @param key
 * @returns
 */
export const safeGetDepsFromPkgJSON = (packageJson: any, key: string) => {
  return packageJson?.dependencies?.[key] || packageJson?.devDependencies?.[key];
};

export const crossPlatformNpm = /^win/.test(process.platform) ? 'npm.cmd' : 'npm';
