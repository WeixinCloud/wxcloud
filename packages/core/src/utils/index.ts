export { logger } from './logger';
export { safeRequire } from './safeRequire';
export { preprocessBaseConfig } from './baseconfig';
