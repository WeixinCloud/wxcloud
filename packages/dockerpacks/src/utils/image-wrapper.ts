import { ImageTag, MatchedImage } from '@api/image';

export class ImageWrapper {
  constructor(private readonly image: MatchedImage) {}

  getFullVersion() {
    return this.image.fullVersion;
  }

  getFullVersionTag() {
    return this.image.fullVersionTag;
  }

  getMostGeneralTag() {
    if (!this.image.matchedTags.length) {
      throw new Error('unexpected matchedTags of image');
    }
    const sorted = [...this.image.matchedTags].sort(sortByGenerality);
    return sorted[0];
  }
}

function sortByGenerality(a: ImageTag, b: ImageTag) {
  switch (true) {
    case !!a.version && !!b.version:
      return countOfDots(a.version!) - countOfDots(b.version!);
    case !a.version && !b.version:
      return 0;
    case !a.version:
      return -1;
    case !b.version:
      return 1;
  }
  throw new Error('unreachable');
}

function countOfDots(input: string) {
  return [...input].filter(c => c === '.').length;
}
