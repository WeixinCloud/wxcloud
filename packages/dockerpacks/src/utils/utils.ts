export function optArg(flag: boolean, value: string | string[]) {
  return flag ? (Array.isArray(value) ? value : [value]) : [];
}
