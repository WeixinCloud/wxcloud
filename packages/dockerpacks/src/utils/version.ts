export const DOCKERPACKS_VERSION = process.env.DOCKERPACKS_VERSION || 'unknown';
