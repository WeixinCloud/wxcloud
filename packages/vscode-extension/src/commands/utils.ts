import * as vscode from 'vscode';
import ext from '../core/global';

interface ICommonCommandHandlingOptions {
  showError?: boolean;
  throwError?: boolean;
  refreshOnFinish?: boolean;
}

export function withCommonCommandHandling(
  fn: (...$args: any[]) => any,
  options: ICommonCommandHandlingOptions = {}
) {
  const { showError = true, throwError = true, refreshOnFinish = true } = options;

  return async (...args: any[]) => {
    try {
      return await fn(...args);
    } catch (e) {
      showError && vscode.window.showErrorMessage(`${e}`);
      if (throwError) {
        throw e;
      }
    } finally {
      refreshOnFinish && ext.wxContainersProvider.refresh();
    }
  };
}
