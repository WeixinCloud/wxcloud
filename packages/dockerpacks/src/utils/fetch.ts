import { merge } from 'lodash';
import fetch, { RequestInfo, RequestInit } from 'node-fetch';
import { DOCKERPACKS_VERSION } from './version';

export function fetchUrl(url: RequestInfo, init?: RequestInit) {
  const source: RequestInit = {
    headers: { 'User-Agent': `Dockerpacks ${DOCKERPACKS_VERSION}` }
  };
  return fetch(url, merge(init ?? {}, source));
}
