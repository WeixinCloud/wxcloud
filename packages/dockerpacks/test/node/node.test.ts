import { describe } from 'vitest';
import { BuilderTestCase, runTest } from '@test/utils';
import path from 'path';

const CASES: BuilderTestCase[] = [
  {
    id: 'build',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'none',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', generalEntrypoint: 'node test/index.js', environments: [] }
  },
  {
    id: 'package-lock',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'npm',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'yarn1',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'yarn2',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'pnpm',
    e2e: { skip: true },
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'create-react-app',
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'create-nuxt-app',
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'create-next-app-yarn',
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'create-next-app-pnpm',
    promptAnswers: { expose: '3000', environments: [] }
  },
  {
    id: 'sveltekit-cli-node-adapter',
    promptAnswers: {
      expose: '3000',
      environments: [],
      generalEntrypoint: 'npm run preview -- --host'
    }
  },
  {
    id: 'sveltekit-cli-static-adapter',
    e2e: { skip: true },
    expectPanic: /adapter-static/i,
    promptAnswers: {
      expose: '3000',
      environments: [],
      generalEntrypoint: 'npm run preview -- --host'
    }
  },
  {
    id: 'expressjs',
    promptAnswers: { expose: '3000', environments: [] }
  }
];

const FIXTURES_PATH = path.join(__dirname, 'fixtures');

describe('builders test (node cases)', () => {
  describe.each(CASES)('case $id', testCase => {
    runTest(FIXTURES_PATH, testCase);
  });
});
