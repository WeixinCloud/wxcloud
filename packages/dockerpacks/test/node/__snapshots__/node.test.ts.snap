// Vitest Snapshot v1

exports[`builders test (node cases) > case build > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 将 package.json 拷贝到容器中
COPY package.json .

# 安装依赖
RUN npm install

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN npm run build

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"start:prod\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case create-next-app-pnpm > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 安装 pnpm
RUN npm install -g pnpm@7.2.1

# 将这些文件拷贝到容器中
COPY package.json pnpm-lock.yaml ./

# 为 pnpm 设置的环境变量
ENV CI=true

# 安装依赖
RUN pnpm install --prefer-offline --shamefully-hoist

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN pnpm run build

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"start\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case create-next-app-yarn > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN yarn config set registry https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN yarn run build

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"start\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case create-nuxt-app > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json package-lock.json ./

# 安装依赖
RUN npm ci

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN npm run build

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"start\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case create-react-app > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json package-lock.json ./

# 安装依赖
RUN npm ci

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN npm run build

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"start\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case expressjs > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN yarn config set registry https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"start\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case none > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"node\\", \\"test/index.js\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md",
  },
}
`;

exports[`builders test (node cases) > case npm > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 安装指定版本的 npm
RUN npm install -g npm@5.10.0

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 将 package.json 拷贝到容器中
COPY package.json .

# 安装依赖
RUN npm install

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"node\\", \\"index.js\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case package-lock > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 安装指定版本的 npm
RUN npm install -g npm@5.10.0

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json package-lock.json ./

# 安装依赖
RUN npm ci

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"node\\", \\"index.js\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case pnpm > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 安装 pnpm
RUN npm install -g pnpm@7.2.1

# 将这些文件拷贝到容器中
COPY package.json pnpm-lock.yml ./

# 为 pnpm 设置的环境变量
ENV CI=true

# 安装依赖
RUN pnpm install --prefer-offline --shamefully-hoist

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"node\\", \\"index.js\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case sveltekit-cli-node-adapter > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN npm config set registry https://registry.npmmirror.com

# 将 package.json 拷贝到容器中
COPY package.json .

# 安装依赖
RUN npm install

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN npm run build

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"npm\\", \\"run\\", \\"preview\\", \\"--\\", \\"--host\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case yarn1 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用速度更快的国内镜像源
RUN yarn config set registry https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"node\\", \\"index.js\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case yarn2 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 使用国内镜像源加速安装 yarn 2.4.3
RUN npm install -g yarn@2.4.3 --registry=\\"https://registry.npmmirror.com\\"

# 使用速度更快的国内镜像源
RUN yarn config set npmRegistryServer https://registry.npmmirror.com

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --immutable

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 设置环境变量
ENV NODE_ENV=production HOST=0.0.0.0

# 运行项目
CMD [\\"node\\", \\"index.js\\"]

# 服务暴露的端口
EXPOSE 3000",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;
